### PR TITLE
PathPlanner interface

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Pushing a refined path planner interface from the A* code review to allow high level logic work around the navigator to be done in parallel with the path planner algorithm itself.

### Testing Done
None. This is just an interface.

### Resolved Issues
#443 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by MathewMacDougall — COMMENTED** [review]
So far LGTM. According to the ticket there should also be a placeholder path planner?

**Comment by jonathanlew** [comment]
Will make placeholder path planner once Jordan approves of interface

**Review by garethellis0 — COMMENTED** [review]
LGTM

File: src/thunderbots/software/ai/navigator/path_planner/path_planner.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+class PathPlanner
+{
+   public:
+    /**
+     * Finds a path from start to dest if it exists, otherwise return std::nullopt.
+     * @param start the start point.
+     * @param dest the destination point.
+     * @return a path from start to dest if it exists, otherwise std::nullopt.
+     */
+    virtual std::optional<std::vector<Point>> findPath(const Point &start,
```
Comment (jordanzliu) [inline src/thunderbots/software/ai/navigator/path_planner/path_planner.h:18]: Some implementations of path planners will require obstacles and/or violation functions. In the case of Theta*, we would use obstacles for graph vertex generation and a violation function for field boundaries. We should provide both of these for the path planner to consume.

**Review by jordanzliu — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/ai/navigator/path_planner/path_planner.h (hunk 1 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/navigator/path_planner/path_planner.h:18]: The violation could be more generic as well. For example if we want to negatively weight certain areas of the field.

Comment (jonathanlew) [reply src/thunderbots/software/ai/navigator/path_planner/path_planner.h:18]: Would A* use violation functions for robots too? Can theta* use boundary polygons for field boundaries?

Comment (garethellis0) [reply src/thunderbots/software/ai/navigator/path_planner/path_planner.h:18]: Yup, at the most abstract level all we need for theta* or A* is a function that tells us if a given point is within an obstacle or not, and a function that gives us a weight for a given point.

**Pull request closed without merging**